### ISSUE TITLE
Training on GPU and predicting on CPU fails.

### ISSUE THREAD

**Issue by roebel** [keras-team/keras#20308]
I am trying to convert code from tf2.14/keras to keras3.5 tf2.16 and things do not work as before.
In my setup I am training on GPU and would like to evaluate on the CPU, which I do notably to measure inference speed on the CPU. I do it like this:

```
with tf.device("GPU"):
    model.compile(...)
    model.fit(...)

with tf.device("CPU"):
   model.predict(...)
```
This did work fine with keras 2.14 and still works fine with legacy mode keras 2.17, but 
fails to run with keras 3.5.

A colab demonstrating the error is [here](https://colab.research.google.com/drive/1oHe2TuBBFQu-ePh9v6lTjFoJJrG0dxQF?usp=sharing)

the error is

```
Trying to access resource dense_model/kernel/5 (defined @ /usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/core.py:40) located in device /job:localhost/replica:0/task:0/device:GPU:0 from device /job:localhost/replica:0/task:0/device:CPU:0
 Cf. https://www.tensorflow.org/xla/known_issues#tfvariable_on_a_different_device
	 [[{{node StatefulPartitionedCall}}]] [Op:__inference_one_step_on_data_distributed_1227]

```
and the related code successfully running with legacy keras is [here](https://colab.research.google.com/drive/1H2IdzVGJrxKXBp3ODrZQCEs5wJ2MN8Ck?usp=sharing).

My question would be whether I am doing something wrong here, or maybe there are alternative approaches to achieving the same thing in keras3. I assume I could save the model and reload it on the CPU, but compared to keras 2 this appears to be a sort of regression. Many thanks for your thoughts.

**Comment by mehtamansi29** [comment]
Hi @roebel - 

Colab shared by you is not accessible for me to reproduce the error. Can you please help me to access those colab to help on this issue ?

**Comment by roebel** [comment]
Hello @mehtamansi29, 
thanks for looking into this. I changed the sharing configuration and you shouid now be able to see the colab. 
Sorry.

**Comment by mehtamansi29** [comment]
Hi @roebel-

Thanks for colab access. Here can train model in keras 3.5.0 using `tf.distribute.MirroredStrategy()`. It will distribute computation across multiple devices. And then predict the on CPU using `keras.device('/device:CPU:0'):`

```
import tensorflow as tf
import keras
import numpy as np

strategy = tf.distribute.MirroredStrategy()
print('Number of devices: {}'.format(strategy.num_replicas_in_sync))

# Define the custom model with a single Dense layer
class DenseModel(keras.Model):
    def __init__(self):
        super().__init__()
        self.dense = keras.layers.Dense(3, activation='relu')
        self.dense2= keras.layers.Dense(1, activation=None)

    def call(self, inputs):
        return self.dense2(self.dense(inputs))

with strategy.scope():
  # Create an instance of the model
  model = DenseModel()

  # Generate some random data for training
  x_train = np.random.random((100, 1))
  y_train = x_train

  # Generate some random data for validation
  x_val = np.random.random((2, 1))
  y_val = x_val
  
  optimizer=keras.optimizers.Adam(learning_rate=0.001)
  # Compile the model
  model.compile(optimizer=optimizer,
                loss='mean_squared_error',
                steps_per_execution=1,)

  # Train the model
  model.fit(x_train, y_train, epochs=2, batch_size=32, validation_data=(x_val, y_val), validation_batch_size=2)

with keras.device('/device:CPU:0'):
  print("predict on batch with steps per execution == 1")
  res = model.predict(x_val)
  print("done")
```

More details regarding `tf.distribute.MirroredStrategy()` can find [here](https://keras.io/guides/distributed_training_with_tensorflow/).

**Comment by mehtamansi29** [comment]
Hi @roebel -

> But what I really wanted to do is running as I always did on keras 2 like [here](https://colab.research.google.com/drive/1TTi5Nn1DQmoVqdZBrBoYK7iTAOjaLMsg?usp=sharing)

As per this colab, the error is like this :
```
Trying to access resource dense_model_4/kernel/89 (defined @ /usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/core.py:40) located in device /job:localhost/replica:0/task:0/device:GPU:0 from device /job:localhost/replica:0/task:0/device:CPU:0
 Cf. https://www.tensorflow.org/xla/known_issues#tfvariable_on_a_different_device
	 [[{{node StatefulPartitionedCall}}]] [Op:__inference_one_step_on_data_distributed_6063]
```
This seems error is related to [XLA](https://openxla.org/xla#tfvariable_on_a_different_device) in tensorflow. XLA compilers takes models from frameworks like(pytorch,tensorflow and jax) and optimize models for high performance execution accross different hardware platforms including GPUs,CPU and ML platforms. 

So keras3.5.0, jit_compile is by default True on GPU. Because of that you are getting XLA related error. 
By changing `jit_compile=False` in model.compile will resolve the error.  `model.compile(optimizer=optimizer,loss='mean_squared_error',steps_per_execution=1,jit_compile=False)`

Attached [gist](https://colab.research.google.com/gist/mehtamansi29/4e6106e40b0dbfc5a433c6d76260ee15/20308-training-on-gpu-and-predicting-on-cpu-fails-1.ipynb) here for your reference. 
Here you can find more related to [jit_compile](https://keras.io/guides/migrating_to_keras_3/#jitcompile-is-set-to-true-by-default-on-gpu) in keras3.

**Comment by roebel** [comment]
Hello mehtamansi29,

Ok, I see. Indeed, running the script with keras 2 and `jit_compile=True` produces the same problem. 
So no regression here!

Following this direction, I have found that I can also avoid the error in my original script by means
of running `model.compile(jit_compile=False)` just before calling  `predict` as follows:

```
model.compile(jit_compile=False)
with keras.device('/device:CPU:0'):
    res = model.predict(x_val)
```
 
So this gives two approaches (strategy.scope() and re-compile) to handle the problem and I guess I should close the issue here.

That said, I am curious to understand why these two approaches work. Given internally for `model.fit` the model uses the `self.train_function`, and for `model.predict` it uses `self.predict_function`, and given these two are compiled independently, I don't see why the predict function tries to use the weights established in the train_function. 

I probably should read [https://www.tensorflow.org/xla/known_issues#tfvariable_on_a_different_device](https://www.tensorflow.org/xla/known_issues#tfvariable_on_a_different_device) but unfortunately this link is dead.
I guess I should pose the question to the tensorflow support

Many thanks for your help!

**Comment by roebel** [comment]
Another solution just for reference, starting with the hypothesis that a keras model is constrained to only contain graphs on a single device this may be the most readable version.

```
# test the model
with keras.device('/device:CPU:0'):
  model_cpu = keras.models.clone_model(model)
  model_cpu.set_weights(model.get_weights())
  print("predict on batch with steps per execution == 1")
  res = model_cpu.predict(x_val)
  print("done")
```
